internal/relationships/validation.go

```diff
@@ -277,3 +299,104 @@ func hasNonEmptyCaveatContext(relationship tuple.Relationship) bool {
 		relationship.OptionalCaveat.Context != nil &&
 		len(relationship.OptionalCaveat.Context.GetFields()) > 0
 }
+
+func CheckDeprecationsOnRelationships(
+	ctx context.Context,
+	relationships []tuple.Relationship,
+	ts *schema.TypeSystem,
+) error {
+	for _, rel := range relationships {
+		if err := checkForDeprecatedRelationsAndObjects(ctx, rel, ts); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func checkForDeprecatedRelationsAndObjects(ctx context.Context, rel tuple.Relationship, ts *schema.TypeSystem) error {
+	// Validate if the resource relation is deprecated
+	relDep, ok, err := ts.GetDeprecationForRelation(ctx, rel.Resource.ObjectType, rel.Resource.Relation)
+	if err != nil {
+		return err
+	}
+	if ok {
+		switch relDep.DeprecationType {
+		case core.DeprecationType_DEPRECATED_TYPE_WARNING:
+			log.Warn().
+				Str("namespace", rel.Resource.ObjectType).
+				Str("relation", rel.Resource.Relation).
+				Str("comments", relDep.Comments).
+				Msg("write to deprecated relation")
+		case core.DeprecationType_DEPRECATED_TYPE_ERROR:
+			return shared.NewDeprecationError(rel.Resource.ObjectType, rel.Resource.Relation, relDep.Comments)
+		}
+	}
+
+	// Validate if the resource namespace is deprecated
+	resDep, ok, err := ts.GetDeprecationForNamespace(ctx, rel.Resource.ObjectType)
+	if err != nil {
+		return err
+	}
+	if ok {
+		switch resDep.DeprecationType {
+		case core.DeprecationType_DEPRECATED_TYPE_WARNING:
+			log.Warn().
+				Str("namespace", rel.Resource.ObjectType).
+				Str("comments", resDep.Comments).
+				Msg("write to deprecated object")
+		case core.DeprecationType_DEPRECATED_TYPE_ERROR:
+			return shared.NewDeprecationError(rel.Resource.ObjectType, "", resDep.Comments)
+		}
+	}
+
+	// Validate if the subject namespace is deprecated
+	subDep, ok, err := ts.GetDeprecationForNamespace(ctx, rel.Subject.ObjectType)
+	if err != nil {
+		return err
+	}
+	if ok {
+		switch subDep.DeprecationType {
```

Comment:
make this "warning vs error" into a helper method and use here and the other two locations